cloud-controller-manager/do/firewall_controller.go

```diff
@@ -117,35 +123,42 @@ func NewFirewallController(
 		workerFirewallName: workerFirewallName,
 		fwManager:          fwManager,
 		metrics:            metrics,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "firewall"),
 	}
 
 	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc: func(cur interface{}) {
-				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
-				defer cancel()
-
-				err := fc.observeReconcileDuration(ctx, originEventAdd)
-				if err != nil {
-					klog.Errorf("failed to reconcile worker firewall after adding a resource object: %s", err)
+				key, err := cache.MetaNamespaceKeyFunc(cur)
+				if err == nil {
+					fc.queue.Add(key)
+
+					reconcileErr := fc.observeReconcileDuration(ctx, originEventAdd)
```

Comment:
If you look at [how the official example works](https://github.com/kubernetes/client-go/blob/ba26c4506d39dece49d00e969d03794e8089b8dc/examples/workqueue/main.go#L175-L194), you'll notice that the `*Func` implementations don't do much anymore other than adding an item to the queue. It is the [`runWorker` loop](https://github.com/kubernetes/client-go/blob/ba26c4506d39dece49d00e969d03794e8089b8dc/examples/workqueue/main.go#L139-L142) which keeps taking items off of that queue and processes them.

(Speaking of, I think we used to have a `runWorker` method at some point but it looks like we lost it. I believe we need to get it back for things to work correctly.)

This means that we need to drop the `observeReconcileDuration` calls from our handlers. The one downside though  is that we lose the information as to which event we processed. I'm not exactly sure if there's a way to pass along context to an item to put onto the queue, we'll need to dig.